include/nbl/builtin/hlsl/emulated_float64_t.hlsl

```diff
@@ -0,0 +1,596 @@
+              zFrac0 = shifted.x;
+              zFrac1 = shifted.y;
+              zFrac2 = shifted.z;
+              zExp = 0;
+              
+              if (roundNearestEven)
+              {
+                 increment = zFrac2 < 0u;
+              }
+              else
+              {
+                 if (zSign != 0u)
+                 {
+                    increment = (FLOAT_ROUNDING_MODE == FLOAT_ROUND_DOWN) && (zFrac2 != 0u);
+                 }
+                 else
+                 {
+                    increment = (FLOAT_ROUNDING_MODE == FLOAT_ROUND_UP) && (zFrac2 != 0u);
+                 }
+              }
+           }
+        
+           if (increment)
+           {
+              nbl::hlsl::uint32_t2 added = add64(zFrac0, zFrac1, 0u, 1u);
+              zFrac0 = added.x;
+              zFrac1 = added.y;
+              zFrac1 &= ~((zFrac2 + uint32_t(zFrac2 == 0u)) & uint32_t(roundNearestEven));
+           }
+           else
+           {
+              zExp = LERP(zExp, 0, (zFrac0 | zFrac1) == 0u);
+           }
+           
+           return packFloat64(zSign, zExp, zFrac0, zFrac1);
+        }
+        
+        uint64_t normalizeRoundAndPackFloat64(uint32_t sign, int exp, uint32_t frac0, uint32_t frac1)
+        {
+           int shiftCount;
+           nbl::hlsl::uint32_t3 frac = nbl::hlsl::uint32_t3(frac0, frac1, 0u);
+        
+           if (frac.x == 0u)
+           {
+              exp -= 32;
+              frac.x = frac.y;
+              frac.y = 0u;
+           }
+        
+           shiftCount = countLeadingZeros32(frac.x) - 11;
+           if (0 <= shiftCount)
+           {
+              frac.xy = shortShift64Left(frac.x, frac.y, shiftCount);
+           }
+           else
+           {
+              frac.xyz = shift64ExtraRightJamming(frac.x, frac.y, 0u, -shiftCount);
+           }
+           exp -= shiftCount;
+           return roundAndPackFloat64(sign, exp, frac.x, frac.y, frac.z);
+        }
+
+    }
+
+    struct emulated_float64_t
+    {
+        using storage_t = uint64_t;
+
+        storage_t data;
+
+        // constructors
+        // TODO: specializations
+        template <typename T>
+        static emulated_float64_t create(T val)
+        {
+#ifndef __HLSL_VERSION
+            emulated_float64_t output;
+            output.data = reinterpret_cast<storage_t&>(val);
+            return output;
+#else
+            uint32_t lowBits;
+            uint32_t highBits;
+            asuint(val, lowBits, highBits);
+            
+            emulated_float64_t output;
+            output.data = (uint64_t(highBits) << 32) | uint64_t(lowBits);
+            return output;
+#endif
+        }
+        
+        static emulated_float64_t createEmulatedFloat64PreserveBitPattern(uint64_t val)
+        {
+            emulated_float64_t output;
+            output.data = val;
+            return output;
+        }

```

Comment:
I think we have our own `bitcast` or `_reinterpret_cast` somewhere that you could specialize for the `T=emulated_float64_t, U=uint64_t` case

Reply:
it will require you define all methods out-of-class though, so you can use that

Reply:
after merging `master` you already have a `bit_cast` we just need to forward it through a struct to be able to do a partial spec